### PR TITLE
Adds BSQL

### PR DESCRIPTION
**Pull request by Cyberboss**
:cl:
server: Added BSQL.dll and updated libmariadb.dll for non-blocking SQL support. See https://github.com/tgstation/BSQL for instructions on building for non-windows systems
/:cl:

BSQL ~uses the [libmariadb non-blocking API](https://mariadb.com/kb/en/library/non-blocking-api-reference)~ runs queries in separate threads to allow us to start queries/connections and check up on them later as opposed to blocking the entire game. SSdbcore has been folded over the BSQL DMAPI so that it works the same as before, only changes are in dbcore.dm. ~~Calls that used to block will now sleep.~~

Eventually we can make the whole stack use an async pattern to cut out the lag. The 4 query execution functions (Execute/warn_execute/NextRow/MassInsert) now include an `async` parameter which will sleep the proc for it's duration instead of blocking.

BSQL.dll was built on release configuration from the v1.3.0.2 tag
libmariadb.dll was built on release configuration from the 3.0.2 tag of https://github.com/MariaDB/mariadb-connector-c (We could potentially get away with not changing this but I'm not familiar with how .libs work in terms of relocation or what version we're currently using)

The exact same binaries can be found here: https://github.com/tgstation/BSQL/releases/tag/v1.3.0.2

TODO:

- [x] Test with tg schema
- [x] Finish writing BSQL README.md
- [x] Test live (HAHAHA GET FUCKED ME)
- [x] Test live more

### CODE REVIEW

**Review by MrsTonedOne — DISMISSED** [review]
This is not going on the servers until all dependencies (outside of libmariadb-client) for building are removed. See the header of the runtime condenser source for an example.

I should be able to clone the git directory. and run cl.exe without visual studio installed, without cmake installed, without anything outside of cl.exe installed.

I don't care if you have to fucking include libmariadb.h in the repo directly to do it.

**Comment by Cyberboss** [comment]
Better dependency handling is doable but doing it without CMake just means I'd have to commit .vcproj project files as well as posix Makefiles in order to get it to build correctly. Any C/C++ developer that doesn't still use autotools will tell you CMake is the one thing necessary to build binaries with more than one source file.

**Comment by MrStonedOne** [comment]
I'll allow it

**Comment by Cyberboss** [comment]
The BSQL readme has been update with command line build instructions

**Comment by Jordie0608** [comment]
With this can we have a way to get error codes and not only the error message?

**Comment by Cyberboss** [comment]
Yeah that's entirely possible, think I'll save it for the next update next time however cause adding functions with the byond garbage api is tedious af

**Comment by Cyberboss** [comment]
Latest commit working in game

**Comment by optimumtact** [comment]
this should have been in rust-g man

**Comment by Cyberboss** [comment]
Didn't because 2 reasons:

1. I hate rust and no one else was stepping up (Be happy it's not in C#)

2. I subscribe to the POSIX philosophy of "does one thing." e.g. Getting /vg/ to adapt another library that overlaps with theirs instead of a single specialized one is more of a challenge. Like all my other projects, this was made to be consumed by the larger BYOND community.

**Comment by MrStonedOne** [comment]
So do we know how well shit handles shit sleeping instead of hard blocking.

I'm pretty sure we are going to have some sort of single legged race condition somewhere from things assuming that no code could ever run between queries.

**Comment by Cyberboss** [comment]
Yeah, we could probably stand to wrap actions that cause multiple inserts/updates (banning? polling?) in transactions

**Comment by MrStonedOne** [comment]
In our game code, is there any code that expects blocking when working with queries. Where they expect no other code could modify the vars they are working in. Do the all recheck for nulls after calling procs that run queries now that they sleep and the vars they once before assume could not change now can change

**Comment by Cyberboss** [comment]
client/New() and world/New() are obvious weakpoints then, not sure how those react to being slept in.

**Comment by MrStonedOne** [comment]
Fine for client, but how about death logging? That likely has issues. Attack procs don't expect sleeping. And admin tools will need null checks now.

**Comment by Cyberboss** [comment]
DNM/Test until the above is merged

**Comment by Cyberboss** [comment]
Gonna do this more atomically, brb

**Comment by Cyberboss** [comment]
Eh, having this as a base is easier

**Comment by Cyberboss** [comment]
Updated OP

**Comment by Cyberboss** [comment]
Def still DNM because of the memory leaks of undeleted queries actually

**Comment by SpaceManiac** [comment]
```
Scanning dependencies of target BSQL
[ 10%] Building CXX object src/BSQL/CMakeFiles/BSQL.dir/BSQL.cpp.o
In file included from /home/travis/build/tgstation/tgstation/BSQL/src/BSQL/BSQL.cpp:1:0:
/home/travis/build/tgstation/tgstation/BSQL/src/BSQL/BSQL.h:7:10: fatal error: mysql/mysql.h: No such file or directory
 #include <mysql/mysql.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/BSQL/CMakeFiles/BSQL.dir/BSQL.cpp.o] Error 1
make[1]: *** [src/BSQL/CMakeFiles/BSQL.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by Cyberboss** [comment]
It's using the exact same shit as BSQL travis, why the hell isn't it working? reee

**Comment by Cyberboss** [comment]
@Jordie0608 I added `/datum/BSQL_Operation/GetErrorCode()` you can have at that once this is merged

**Comment by Cyberboss** [comment]
Travis is fixed up. This is in a testable state now.

File: code/controllers/subsystem/dbcore.dm
Comment (MrStonedOne) [inline code/controllers/subsystem/dbcore.dm]: it doesn't appear you use this parameter

**Comment by Cyberboss** [comment]
Fixed

File: code/controllers/subsystem/dbcore.dm
Comment (MrStonedOne) [inline code/controllers/subsystem/dbcore.dm]: I was going to suggest making this asyncable but i suppose the primary issue (long&blocking connection timeouts) could be fixed by just using a sane connection timeout. what is bsql using?

Comment (Cyberboss) [reply code/controllers/subsystem/dbcore.dm]: uhh whatever the default is for mariadb. (I've been meaning to add that as a parameter, I'll get on that)

File: config/dbconfig.txt
Comment (Cyberboss) [inline config/dbconfig.txt]: @MrStonedOne

Comment (Cyberboss) [reply config/dbconfig.txt]: If the timeout elapses it runs the current slow query handler

**Comment by Cyberboss** [comment]
Merge this tomorrow if no one complains

File: config/dbconfig.txt
Comment (MrStonedOne) [reply config/dbconfig.txt]: Re read it again

Comment (MrStonedOne) [reply config/dbconfig.txt]: word _for_ word

Comment (optimumtact) [reply config/dbconfig.txt]: communicate clearly

Comment (MrStonedOne) [reply config/dbconfig.txt]: read it word _for_ word

or rather

read it _for_ word word.

Comment (ike709) [reply config/dbconfig.txt]: >Time in seconds for blocking queries to execute before alerting a for possible slow query timeout 

Read it slowly.

Comment (optimumtact) [reply config/dbconfig.txt]: you can just say, there's an extra a before your for in this sentence

you don't have to play 20 questions, we're not on the chaser

Comment (MrStonedOne) [reply config/dbconfig.txt]: but there isn't.

Comment (MrStonedOne) [reply config/dbconfig.txt]: the isn't an extra `a`.

Comment (Cyberboss) [reply config/dbconfig.txt]: #39173

Comment (optimumtact) [reply config/dbconfig.txt]: there is an extra a

Comment (MrStonedOne) [reply config/dbconfig.txt]: no, there is exactly the right amount of a's in that sentence, its just that one of them appears 1 word too soon.

File: code/controllers/subsystem/dbcore.dm
Comment (JohnGinnane) [reply code/controllers/subsystem/dbcore.dm]: why is this false?

**Pull request merged**
